docs/design.md

```diff
@@ -0,0 +1,258 @@
+# Design
+
+This document explains the purpose of the `sops-compliance-checker` and also
+documents the functional and non-functional requirements we have, as well as
+the details about design decisions we took. It assumes familarity with
+[SOPS][sops] and the concepts that it uses.
+
+## Background
+
+SOPS supports encrypting files using one or more trust anchors. Especially in
+enterprise contexts, it is important to enforce some rules around the trust
+anchors that are in use. Some questions that to be answered could
+be:
+
+- How can we ensure that any given SOPS file is encrypted using a set of
+  required trust anchors?
+- How can we detect if non-approved trust anchors are used?
+- How can we prevent the usage of certain trust anchors in the presence of
+  others (mutual exclusivity)?
+
+To answer these questions for any given SOPS file, the
+`sops-compliance-checker` needs to be flexible enough to process a set of
+user-defined rules which indicate success or failure.
+
+## Functional Requirements
+
+The high-level functionality of the compliance checker can be summarized like
+this:
+
+- Recursively scan a directory tree (such as a Git repository) for SOPS
+  encrypted files.
+- It should support all file formats supported by SOPS itself.
+- Extract trust anchors from SOPS encrypted files to match them against a set
+  of user-defined rules.
+- Report check results back to the user.
+
+The rules engine needs to support the following functionality:
+
+- Matching of a trust anchor ("match")
+  - **Example**: The trust anchor is an AWS KMS Key with the ARN
+    `arn:aws:kms:eu-west-1:123456789012:alias/my-team`.
```

Comment:
I think we should also support partial matches via **pattern** or **regex**, so that we can for example check if a KMS key belongs to our AWS account.

Reply:
I actually intended to support pattern matching via regex and it is even mentioned in the proposed schema: https://github.com/Bonial-International-GmbH/sops-compliance-checker/blob/60292a71c22683932f56820dced6b65f5e31c038/schema.json#L36

I added another matching example for regular expressions now to make this more obvious.